### PR TITLE
Change TokenSource to iteration based

### PR DESCRIPTION
**Pull request by edwin0cheng**
This PR change the `TokenSource` trait from random access to be an iteration based trait:
```rust

/// `TokenSource` abstracts the source of the tokens parser operates one.
///
/// Hopefully this will allow us to treat text and token trees in the same way!
pub trait TokenSource {
    fn current(&self) -> Token;

    /// Lookahead n token
    fn lookahead_nth(&self, n: usize) -> Token;

    /// bump cursor to next token
    fn bump(&mut self);

    /// Is the current token a specified keyword?
    fn is_keyword(&self, kw: &str) -> bool;
}

/// `TokenCursor` abstracts the cursor of `TokenSource` operates one.
#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct Token {
    /// What is the current token?
    pub kind: SyntaxKind,

    /// Is the current token joined to the next one (`> >` vs `>>`).
    pub is_jointed_to_next: bool,
}
```

Note that the refactoring based on this new trait will be separated to incoming PRs

### CODE REVIEW

File: crates/ra_mbe/src/subtree_source.rs [hunk 1]
```diff
@@ -117,43 +118,59 @@ impl<'a> Querier for SubtreeWalk<'a> {
...
+        self.curr.0
     }
-    fn is_token_joint_to_next(&self, pos: usize) -> bool {
-        match self.walker.get(pos) {
-            Some(t) => t.is_joint_to_next,
-            _ => false,
+
+    /// Lookahead n token
+    fn lookahead_nth(&self, n: usize) -> Token {
+        self.mk_token(self.curr.1 + n)
```
Comment (edwin0cheng) [inline crates/ra_mbe/src/subtree_source.rs]: Originally it should be `lookahead_1` , `lookahead_2`, `lookahead_3` , but compositable tokens and DOLLARS makes it impossible.

File: crates/ra_parser/src/lib.rs [hunk 2]
```diff
@@ -31,12 +31,26 @@ pub struct ParseError(pub String);
 ///
 /// Hopefully this will allow us to treat text and token trees in the same way!
 pub trait TokenSource {
+    fn current(&self) -> Token;
+
+    /// Lookahead n token
+    fn lookahead_nth(&self, n: usize) -> Token;
```
Comment (edwin0cheng) [inline crates/ra_parser/src/lib.rs]: Originally it should be `lookahead_1` , `lookahead_2`, `lookahead_3` , but composite-able tokens and DOLLARS makes it impossible.

File: crates/ra_mbe/src/syntax_bridge.rs [hunk 3]
```diff
@@ -121,9 +126,10 @@ pub fn token_tree_to_macro_items(
 /// Parses the token tree (result of macro expansion) as a sequence of items
 pub fn token_tree_to_ast_item_list(tt: &tt::Subtree) -> TreeArc<ast::SourceFile> {
     let buffer = tt::buffer::TokenBuffer::new(&[tt.clone().into()]);
-    let token_source = SubtreeTokenSource::new(&buffer);
-    let mut tree_sink = TtTreeSink::new(token_source.querier());
-    ra_parser::parse(&token_source, &mut tree_sink);
+    let mut token_source = SubtreeTokenSource::new(&buffer);
+    let querier = token_source.querier();
+    let mut tree_sink = TtTreeSink::new(querier.as_ref());
+    ra_parser::parse(&mut token_source, &mut tree_sink);
```
Comment (matklad) [inline crates/ra_mbe/src/syntax_bridge.rs]: I wonder if these 6 lines could be extracted into a helper function?

Comment (edwin0cheng) [reply crates/ra_mbe/src/syntax_bridge.rs]: Fixed.

File: crates/ra_mbe/src/syntax_bridge.rs [hunk 4]
```diff
@@ -45,92 +45,65 @@ pub fn syntax_node_to_token_tree(node: &SyntaxNode) -> Option<(tt::Subtree, Toke
...
+where
+    F: Fn(&mut ra_parser::TokenSource, &mut ra_parser::TreeSink),
+{
     let buffer = tt::buffer::TokenBuffer::new(&[tt.clone().into()]);
     let mut token_source = SubtreeTokenSource::new(&buffer);
     let querier = token_source.querier();
     let mut tree_sink = TtTreeSink::new(querier.as_ref());
-    ra_parser::parse_expr(&mut token_source, &mut tree_sink);
+    f(&mut token_source, &mut tree_sink);
+    if tree_sink.roots.len() != 1 {
```
Comment (andreevlex) [inline crates/ra_mbe/src/syntax_bridge.rs]: The repetition of code. Below is such a code there

Comment (edwin0cheng) [reply crates/ra_mbe/src/syntax_bridge.rs]: Good catch, fixed!

**Pull request merged**